src/raiden_libs/matrix.py

```diff
@@ -374,3 +271,170 @@ def _handle_message(self, room: Room, message: MatrixMessage) -> List[SignedMess
+
+        self.main_client = make_client(
+            handle_messages_callback=handle_matrix_sync,
+            handle_member_join_callback=lambda room: None,
+            servers=self.available_servers,
+            http_pool_maxsize=4,
+            http_retry_timeout=40,
+            http_retry_delay=matrix_http_retry_delay,
+        )
+        self.server_url_to_other_clients: Dict[str, GMatrixClient] = {}
+        self.connect_client_workers: Set[Greenlet] = set()
+
+    @property
+    def server_url_to_all_clients(self) -> Dict[str, GMatrixClient]:
+        return {
+            **self.server_url_to_other_clients,
+            urlparse(self.main_client.api.base_url).netloc: self.main_client,
+        }
+
+    def start(self, user_manager: MultiClientUserAddressManager) -> None:
+        self.stop_event.clear()
+        self.user_manager = user_manager
+        try:
+            self._start_client(self.main_client.api.base_url)
+        except (TransportError, ConnectionError):
+            # When the sync worker fails, waiting for startup_finished does not
+            # make any sense.
+            self.startup_finished.set()
+            return
+
+        for server_url in [
+            server_url
+            for server_url in self.known_servers
+            if server_url != self.main_client.api.base_url
+        ]:
+            connect_worker = gevent.spawn(self.connect_client_forever, server_url)
+            self.connect_client_workers.add(connect_worker)
+
+    def stop(self) -> None:
+        assert self.user_manager, "Stop called before start"
+
+        self.stop_event.set()
+        for server_url, client in self.server_url_to_all_clients.items():
+            self.server_url_to_other_clients.pop(server_url, None)
+            client.stop_listener_thread()
+            self.user_manager.remove_client(client)
+
+        gevent.joinall(self.connect_client_workers, raise_error=True)
+
+    def connect_client_forever(self, server_url: str) -> None:
+        assert self.user_manager
+        while not self.stop_event.is_set():
+            stopped_client = self.server_url_to_other_clients.pop(server_url, None)
+            if stopped_client is not None:
+                self.user_manager.remove_client(stopped_client)
+            try:
+                client = self._start_client(server_url)
+                assert client.sync_worker is not None
+                client.sync_worker.get()
+            except (TransportError, ConnectionError):
+                log.debug("Could not connect to server", server_url=server_url)
+
+    def _start_client(self, server_url: str) -> GMatrixClient:
+        assert self.user_manager
+        if self.stop_event.is_set():
+            raise TransportError()
+
+        if server_url == self.main_client.api.base_url:
+            client = self.main_client
+        else:
+            client = make_client(
+                handle_messages_callback=lambda messages: True,
```

Comment:
In order for the other clients to also listen on toDevice messages you need to pass the handler here as well

Reply:
The newest commit now passes the main-client's handler to all sub-clients.
Note that the message handler accepts to-device as well as room-messages for all clients.
If duplicated message handling at this point is a problem, the handlers would have to be separated,
so that only the main client listens on the room.